syntax = "proto3";

package synapse;

import "api/datatype.proto";
import "api/nodes/electrical_broadband.proto";
import "api/nodes/electrical_stim.proto";
import "api/nodes/optical_broadband.proto";
import "api/nodes/optical_stim.proto";
import "api/nodes/spike_detect.proto";
import "api/nodes/stream_out.proto";
import "api/nodes/stream_in.proto";

enum NodeType {
  kNodeTypeUnknown = 0;
  kStreamIn = 1;
  kStreamOut = 2;
  kElectricalBroadband = 3;
  kElectricalStim = 4;
  kOpticalBroadband = 5;
  kOpticalStim = 6;
  kSpikeDetect = 7;
  kSpectralFilter = 8;
}



message NodeOptions {
  NodeType type = 1;
  uint32 id = 2;
  oneof options {
    ElectricalBroadbandOptions electrical_broadband = 3;
    ElectricalStimOptions electrical_stim = 4;
    OpticalBroadbandOptions optical_broadband = 5;
    OpticalStimOptions optical_stim = 6;
    SpikeDetectOptions spike_detect = 7;
  }
}

message NodeConfig {
  NodeType type = 1;
  uint32 id = 2;
  oneof config {
    ElectricalBroadbandConfig electrical_broadband = 3;
    ElectricalStimConfig electrical_stim = 4;
    OpticalBroadbandConfig optical_broadband = 5;
    OpticalStimConfig optical_stim = 6;
    SpikeDetectConfig spike_detect = 7;
    StreamOutConfig stream_out = 8;
    StreamInConfig stream_in = 9;
  }
}

message NodeConnection {
  uint32 src_node_id = 1;
  uint32 dst_node_id = 2;
}

message NodeSocket {
  uint32 node_id = 1;
  uint32 bind = 2;
  DataType data_type = 3;
  NodeType type = 4;
}